### ISSUE TITLE
Surprising `error: [mustcall:methodref.receiver] Incompatible receiver type`

### ISSUE THREAD

**Issue by jacek-lewandowski** [typetools/checker-framework#6038]
The analyser complains about this method:

```java
    private static <T> List<String> toStringList(Collection<T> list)
    {
        if (list == null)
            return null;
        return list.stream().map(Object::toString).collect(Collectors.toList());
    }
```

with

```
    [javac] /home/jlewandowski/dev/cassandra/c18190-ecj/src/java/org/apache/cassandra/db/virtual/LocalRepairTables.java:460: error: [mustcall:methodref.receiver] Incompatible receiver type
    [javac]         return list.stream().map(Object::toString).collect(Collectors.toList());
    [javac]                                  ^
    [javac]   found   : @MustCall Object
    [javac]   required: T[ extends @MustCallUnknown Object super @MustCall Void]
    [javac]   Consequence: method in @MustCall Object
    [javac]     @MustCall String toString(@MustCall Object this)
    [javac]   is not a valid method reference for method in @MustCallUnknown Function<T[ extends @MustCallUnknown Object super @MustCall Void], @MustCall String>
    [javac]     @MustCall String apply(@MustCall Function<T[ extends @MustCallUnknown Object super @MustCall Void], @MustCall String> this, T[ extends @MustCallUnknown Object super @MustCall Void] p0)
```

I must admit I do not understand why it complains :)

**Comment by jacek-lewandowski** [comment]
Another example is:

```java
    /**
     * Constuct a textual representation of the trie.
     */
    public String dump()
    {
        return dump(Object::toString);
    }

    /**
     * Constuct a textual representation of the trie using the given content-to-string mapper.
     */
    public String dump(Function<T, String> contentToString)
    {
        return process(new TrieDumper<>(contentToString));
    }
```

```
    [javac] /home/jlewandowski/dev/cassandra/c18190-ecj/src/java/org/apache/cassandra/db/tries/Trie.java:362: error: [mustcall:methodref.receiver] Incompatible receiver type
    [javac]         return dump(Object::toString);
    [javac]                     ^
    [javac]   found   : @MustCall Object
    [javac]   required: T[ extends @MustCallUnknown Object super @MustCall Void]
    [javac]   Consequence: method in @MustCall Object
    [javac]     @MustCall String toString(@MustCall Object this)
    [javac]   is not a valid method reference for method in @MustCallUnknown Function<T[ extends @MustCallUnknown Object super @MustCall Void], @MustCall String>
    [javac]     @MustCall String apply(@MustCall Function<T[ extends @MustCallUnknown Object super @MustCall Void], @MustCall String> this, T[ extends @MustCallUnknown Object super @MustCall Void] p0)
```

What I have noticed is that, if I use a lambda expression instead of a method reference, it passes the analysis without problems.

**Comment by kelloggm** [comment]
These warnings are a consequence of imprecision in the analysis when analyzing method references. They can safely be suppressed. It would be nice if the RLC didn't warn in these cases, but it's technically difficult to achieve, they aren't particularly common, and it's easy to tell that they are false positives (and can therefore safely be suppressed).

In particular, here's what I'd do with your first example (note the new `@SuppressWarnings` annotation):
```
    @SuppressWarnings("mustcall: methodref.receiver") // false positive due to imprecision: https://github.com/typetools/checker-framework/issues/6038
    private static <T> List<String> toStringList(Collection<T> list)
    {
        if (list == null)
            return null;
        return list.stream().map(Object::toString).collect(Collectors.toList());
    }

```

**Comment by jacek-lewandowski** [comment]
thanks

**Comment by kelloggm** [comment]
I'd prefer to leave this issue open, because these warnings are false positives and it would be great to fix them if we can figure out an easy way to do so.

**Comment by kelloggm** [comment]
That said, this is probably wontfix
